system/lib/pthread/proxying.h

```diff
@@ -67,5 +67,82 @@ int emscripten_proxy_sync_with_ctx(em_proxying_queue* q,
                                    void* arg);
 
 #ifdef __cplusplus
-}
-#endif
+} // extern "C"
+
+#if __cplusplus < 201103L
+#warning "C++ ProxyingQueue support requires building with -std=c++11 or newer!"
+#else
+
+#include <functional>
+#include <thread>
+#include <utility>
+
+namespace emscripten {
+
+// A thin C++ wrapper around the underlying C API.
+class ProxyingQueue {
+  em_proxying_queue* queue = nullptr;
+
+  static void runAndFree(void* arg) {
+    auto f = (std::function<void()>*)arg;
+    (*f)();
+    delete f;
+  }
+
+  static void run(void* arg) {
+    auto f = *(std::function<void()>*)arg;
+    f();
+  }
+
+  static void runWithCtx(em_proxying_ctx* ctx, void* arg) {
+    auto f = *(std::function<void(ProxyingCtx)>*)arg;
+    f(ProxyingCtx{ctx});
+  }
+
+public:
+  ProxyingQueue() { queue = em_proxying_queue_create(); }
```

Comment:
Can you just do this in the member declaration these days? e.g.:

`em_proxying_queue* queue = em_proxying_queue_create()` above?